# EXOTIC FRUIT CLASSIFICATION 
*Fourth Master AI Engineering project*

## Import Libraries and Dataset

* The Python packages I used in this projects are in requirements.txt file
* Dataset is stored in the 'Data' folder
* I renamed the columns for two reasons:
    - I need to translate Italian column names into English ones
    - I didn't want measurement units and ranges in the column names

In [1]:
# Import libraries

import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns
import plotly.express as px
import pygwalker as pyg

from src.preprocessing_and_viz import handle_duplicates, plot_feature_distribution, scatter_plot, feature_scaling

In [2]:
# Import Dataset

df = pd.read_csv("Data/Fruits_Dataset.csv")

df.head()

Frutto  Peso (g)  Diametro medio (mm)  Lunghezza media (mm)  \
0   Mela     86.40                89.68                  8.69   
1   Mela     77.58                73.45                  6.49   
2   Mela     81.95                81.66                  6.40   
3   Mela     66.33                36.71                  6.78   
4   Mela     56.73                75.69                  5.78   

   Durezza buccia (1-10)  Dolcezza (1-10)  
0                   9.61             2.41  
1                   7.20             3.87  
2                   9.09             2.88  
3                   8.21             2.55  
4                   9.15             3.88

In [3]:
# Renaming columns

new_columns = [
    'Fruit',
    'Weight',
    'Average diameter',
    'Average length',
    'Peel hardness',
    'Sweetness',
]

df.columns = new_columns

df.columns

Index(['Fruit', 'Weight', 'Average diameter', 'Average length',
       'Peel hardness', 'Sweetness'],
      dtype='object')

## Statistics Information

Useful information I got:
* Shape: (500,6)
* `Fruit` is the target column. There are 5 options: Apple, Banana, Orange, Grape, Kiwi
* 5 numerical features with continous values (float64)
* Statistics:
    - mean values and ranges are one order of magnitude different, I will implement normalization/standardization
    - Maximum value for Peel hardness should have been 10, but it's clearly higher (13.72). Strange!
    - it seems there are no missing values 

In [7]:
# Shape and Info

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Fruit             500 non-null    object 
 1   Weight            500 non-null    float64
 2   Average diameter  500 non-null    float64
 3   Average length    500 non-null    float64
 4   Peel hardness     500 non-null    float64
 5   Sweetness         500 non-null    float64
dtypes: float64(5), object(1)
memory usage: 23.6+ KB


In [8]:
# Statistics

df.describe()

Weight  Average diameter  Average length  Peel hardness   Sweetness
count  500.000000        500.000000      500.000000     500.000000  500.000000
mean    47.313440         86.573100        5.298620       7.437640    3.593220
std     26.768797         64.293403        2.641993       1.812548    1.264899
min      8.570000          7.530000        1.150000       3.070000    1.250000
25%     24.797500         50.510000        2.672500       6.152500    2.570000
50%     42.380000         70.450000        5.670000       7.340000    3.535000
75%     68.080000         88.852500        7.455000       8.615000    4.465000
max    111.210000        299.890000       11.140000      13.720000    6.950000

In [4]:
# Rename Target values and see how they are distributed

target_value_names = {
    'Mela' : 'Apple',
    'Banana' : 'Banana',
    'Arancia' : 'Orange',
    'Uva' : 'Grape',
    'Kiwi' : 'Kiwi'
}

df['Fruit'] = df['Fruit'].map(lambda x: target_value_names[x])

df['Fruit'].value_counts()

Fruit
Apple     100
Banana    100
Orange    100
Grape     100
Kiwi      100
Name: count, dtype: int64

## Preprocessing

* No duplicates in the dataset. Shape is still (500,6)
* No missing values in the dataset.

In [5]:
# Check for duplicates

df_cleaned = handle_duplicates(df)

No duplicates found!


In [8]:
# Check for missing values

df_cleaned.isna().sum()

Fruit               0
Weight              0
Average diameter    0
Average length      0
Peel hardness       0
Sweetness           0
dtype: int64

## Data Visualization

Feature distribution without comparison with target column:
* Weight: as expected, it is a negatively skewed distribution with high max value (111.21)
* Average diameter: it is clear we have smaller fruits, so distribution is negatively skewed, with higher values come from Banana target
* Average length: its distribution has two peaks, sign of fruits with well separated lengths
* Peel hardness: has a almost normal distribution with a couple of possible outliers, median and mean are quite the same
* Sweetness: has a almost normal distribution, median and mean are quite the same

Feature distribution and scatter plot and comparison with target column:
* Weight, Average diameter, Average length: each fruit follows distributions that were expected, interesting how Apple, Orange and Kiwi have almost the same distribution for Average diameter
* Peel hardness: each fruit seems to follow a normal distribution
* Sweetness: Orange, Grape and Kiwi reach higher values, as expected

Scatter plot to see how features are distributed according to target column:
* Weight vs Average diameter: it is possible to clearly see two clusters for Banana and Grape. Maybe, Apple, Orange and Kiwi will be more difficult to classify
* Average length vs Average diameter: this plot emphasizes what said before 
* Peel hardness vs Weight: lighter fruits have higher Peel hardness values
* Sweetness vs Average diameter: smaller fruits have higher Sweetness values
* Peel hardness vs Sweetness: Banana has lowest Sweetness values and well distributed Peel hardness values. Orange, instead, had higher Sweetness values and lowest Peel hardness values

In [28]:
# Feature distribution among ['Weight', 'Average diameter', 'Average length', 'Peel hardness', 'Sweetness']

feature_to_viz = 'Weight'

plot_feature_distribution(df_cleaned, feature_to_viz, comparison=False)

In [33]:
# Feature distribution among ['Weight', 'Average diameter', 'Average length', 'Peel hardness', 'Sweetness']
# and comparison with TARGET 'Fruit' 

feature_to_viz = 'Sweetness'

plot_feature_distribution(df_cleaned, feature_to_viz, comparison='Fruit')

In [39]:
# Scatter plot among ['Weight', 'Average diameter', 'Average length', 'Peel hardness', 'Sweetness']
# and comparison with TARGET 'Fruit' 

feature_x = 'Sweetness'
feature_y = 'Average diameter'

scatter_plot(df_cleaned, feature_x, feature_y, target='Fruit')

In [ ]:
# Visualization Dashboard using PYGWALKER, you can see interactively the last plots
# You can save the json file for the configuration you want to visualize and make other people see it using this code:
# pyg.walk(df_cleaned, spec="Images/pygwalker_config_1.json")


pyg.walk(df_cleaned)

## Feature Engineering 

In [7]:
# Feature scaling with MinMaxScaler method

df_transformed = feature_scaling(df_cleaned, mixed = False)
df_transformed.describe()

Weight  Average diameter  Average length  Peel hardness   Sweetness
count  500.000000        500.000000      500.000000     500.000000  500.000000
mean     0.377469          0.270362        0.415277       0.410107    0.411091
std      0.260803          0.219912        0.264464       0.170192    0.221912
min      0.000000          0.000000        0.000000       0.000000    0.000000
25%      0.158101          0.147011        0.152402       0.289437    0.231579
50%      0.329404          0.215214        0.452452       0.400939    0.400877
75%      0.579793          0.278159        0.631131       0.520657    0.564035
max      1.000000          1.000000        1.000000       1.000000    1.000000

In [8]:
# Feature scaling with MinMaxScaler method for NOT normal distributed feature, StandardScaler method for normal distributed features

df_mix_transformed = feature_scaling(df_cleaned, mixed = True)
df_mix_transformed.describe()

Weight  Average diameter  Average length  Peel hardness  \
count  500.000000        500.000000      500.000000   5.000000e+02   
mean     0.377469          0.270362        0.415277   4.902745e-16   
std      0.260803          0.219912        0.264464   1.001002e+00   
min      0.000000          0.000000        0.000000  -2.412082e+00   
25%      0.158101          0.147011        0.152402  -7.097340e-01   
50%      0.329404          0.215214        0.452452  -5.392286e-02   
75%      0.579793          0.278159        0.631131   6.502112e-01   
max      1.000000          1.000000        1.000000   3.469509e+00   

          Sweetness  
count  5.000000e+02  
mean  -2.842171e-17  
std    1.001002e+00  
min   -1.854352e+00  
25%   -8.097445e-01  
50%   -4.607350e-02  
75%    6.898996e-01  
max    2.656451e+00

## Visualization after feature engineering

In [9]:
# Scatter plot among ['Weight', 'Average diameter', 'Average length', 'Peel hardness', 'Sweetness']
# and comparison with TARGET 'Fruit' 

feature_x = 'Sweetness'
feature_y = 'Average diameter'

scatter_plot(df_transformed, feature_x, feature_y, target='Fruit')

##  K-Nearest Neighbors (KNN) implementation

## Results

## [EXTRA] Machine Learning Models implementation